internal/clone/registry.go

```diff
@@ -0,0 +1,161 @@
+// using under those interfaces. This registration needs to happen on both sides i.e the
+// side which encodes the data and the side which decodes the data.
+// Go init function: https://golang.org/doc/effective_go.html#init
+func init() {
+	// Register the pointer to structs because that is what is being stored.
+	gob.Register(&uvm.VSMBShare{})
+	gob.Register(&uvm.SCSIMount{})
+}
+
+func encodeTemplateConfig(utc *uvm.UVMTemplateConfig) ([]byte, error) {
+	var buf bytes.Buffer
+
+	encoder := gob.NewEncoder(&buf)
+	if err := encoder.Encode(utc); err != nil {
+		return nil, fmt.Errorf("error while encoding template config: %s", err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeTemplateConfig(encodedBytes []byte) (*uvm.UVMTemplateConfig, error) {
+	var utc uvm.UVMTemplateConfig
+
+	reader := bytes.NewReader(encodedBytes)
+	decoder := gob.NewDecoder(reader)
+	if err := decoder.Decode(&utc); err != nil {
+		return nil, fmt.Errorf("error while decoding template config: %s", err)
+	}
+	return &utc, nil
+}
+
+// loadPersistedConfig loads a persisted config from the registry that matches the given ID
+// If not found returns `regstate.NotFoundError`
+func loadPersistedUVMConfig(ID string) (*persistedUVMConfig, error) {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		return nil, err
+	}
+	defer sk.Close()
+
+	puc := &persistedUVMConfig{
+		Stored: true,
+	}
+	if err := sk.Get(ID, configKey, puc); err != nil {
+		return nil, err
+	}
+	return puc, nil
+}
+
+// storePersistedUVMConfig stores or updates the in-memory config to its registry state.
+// If the store fails returns the store error.
+func storePersistedUVMConfig(ID string, puc *persistedUVMConfig) error {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		return err
+	}
+	defer sk.Close()
+
+	if puc.Stored {
+		if err := sk.Set(ID, configKey, puc); err != nil {
+			return err
+		}
+	} else {
+		if err := sk.Create(ID, configKey, puc); err != nil {
+			return err
+		}
+	}
+	puc.Stored = true
+	return nil
+}
+
+// removePersistedUVMConfig removes any persisted state associated with this config. If the config
+// is not found in the registery `Remove` returns no error.
+func removePersistedUVMConfig(ID string) error {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		if regstate.IsNotFoundError(err) {
+			return nil
+		}
+		return err
+	}
+	defer sk.Close()
+
+	if err := sk.Remove(ID); err != nil {
+		if regstate.IsNotFoundError(err) {
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+// Saves all the information required to create a clone from the template
+// of this container into the registry.
+func SaveTemplateConfig(ctx context.Context, utc *uvm.UVMTemplateConfig) error {
+	_, err := loadPersistedUVMConfig(utc.UVMID)
+	if !regstate.IsNotFoundError(err) {
+		return fmt.Errorf("parent VM(ID: %s) config shouldn't exit in registry (%s) \n", utc.UVMID, err.Error())
```

Comment:
Don't need to call `.Error()` if the format string is `%s`. Also, Go error strings shouldn't end in a newline.